### ISSUE TITLE
sql: Full Text Search

### ISSUE THREAD

**Issue by Linicks** [cockroachdb/cockroach#7821]
All,
    I'm sure most of you know about Bleve (https://github.com/blevesearch/bleve)  a Go-lang based full-text indexer.  I was wondering if you've considered integrating it with CockroachDB?  I'ts seems like it may be a good fit, and is being used in other distributed databases.  
-  It has an Apache licence.
- Written in Go.
- Already provides allot of value add functionality.

Thanks!
-- Nick 

*Maintainer note from @jordanlewis: see the following issues for our current progress on search*

- [x] Trigram indexing support: #41285
- [x] tsvector / tsquery support: #41288

gz#6861

Jira issue: CRDB-6169

**Comment by petermattis** [comment]
@Linicks Full-text search is something we'd like to support and Bleve is on my radar, though there are no concrete plans to integrate it.

**Comment by alexander-manley** [comment]
One approach for integrating Bleve with Cockroach, and thus provide CockroachDB with text search, would be to modify hugoidx (https://github.com/blevesearch/hugoidx) to allow it to BLEVE-index the contents of a Cockroach BLOB store (...https://github.com/cockroachdb/cockroach/issues/243) pre-populated with corpus text (web page scrapes, text-doc-dumps etc...).

In addition to **hugoidx**, the associated Go utility **"bleve-hosted"** could be wrapped into the embedded UI (https://github.com/cockroachdb/cockroach/tree/master/ui) in order to pull-out and/or highlight text search results pulled from the BLOB store and displayed as an additional panel under the left side "DATABASES" UI tab.

Bleve is based on file indexes, which by default are stored in BoltDB, so that part would need to be ported over to RocksDB for full integration.  For the curious, a Bleve benchmark graph with RocksDB was posted to the  Bleve Twitter stream a while back.

Reference: 
http://www.blevesearch.com/news/Site-Search/
http://www.blevesearch.com/videos/

**Comment by petermattis** [comment]
@alexander-manley Thanks for the notes. We'll definitely take a closer look at Bleve when considering full-text indexing.

**Comment by randyyaj** [comment]
Any updates on this?

**Comment by petermattis** [comment]
@randyyaj Full-text indexing is something we'd like to do, but still a ways off and not currently scheduled.

**Comment by SantoshSah** [comment]
@petermattis , any update?

**Comment by nstewart** [comment]
Full text search is something we want to support, but it is not on the roadmap for cockroachdb 2.1 or 2.2. While we are adding some new functionality, for the next couple releases we are focusing on improving the performance and stability of our current offering before we add major new features.

**Comment by RoachietheSupportRoach** [comment]
Zendesk ticket #[3521](https://cockroachdb.zendesk.com/agent/tickets/3521) has been linked to this issue.

**Comment by OldhamMade** [comment]
Does this zendesk ticket mean that full-text indexing is being actively worked on?

**Comment by jordanlewis** [comment]
No, full text search isn't on the near term roadmap for the time being.

>

**Comment by jezell** [comment]
I think the best way to get some sort of support for full text search "out of the box" would be to CREATE CHANGEFEED to support some destinations like elasticsearch, vespa, algolia, etc. Modern full text search is a completely different domain than relational data. While I'm sure the team could eventually crack it, it would likely be a long road to get it up to par with something like Vespa. I'd personally rather see out of the box integration, as we wouldn't want to give up search result quality to switch to something built in.

**Comment by jordanlewis** [comment]
CockroachDB 22.2 will support trigram indexes, a simple form of text search that may help some of your use cases. See #79705 for details on what has been added.

**Comment by amirouche** [comment]
Since I only used it for spell checking for small dictionaries, I am not sure how trigrams help to implement full-text search.

**Comment by jordanlewis** [comment]
Please feel free to follow and upvote #41288, which is an issue that tracks Postgres-compatible tsvector and tsquery implementations.
